net/dns: retry forwarder requests over TCP

andrew-d opened the pull request:
We weren't correctly retrying truncated requests to an upstream DNS server with TCP. Instead, we'd return a truncated request to the user, even if the user was querying us over TCP and thus able to handle a large response.

Updates #9264


Change-Id: Ifb04b563839a9614c0ba03e9c564e8924c1a2bfd

andrew-d commented:
Marking as draft since I think there's one edge case this doesn't handle: the incoming request is UDP, we retry an upstream query over TCP and get a response, but then need to send it back over UDP. Do we need to truncate the response at that point?

twitchyliquid64 commented:
Im ... surprised we didnt implement looking up DNS over TCP lol.

WRT the edge case: yes. If the request we are forwarding is over UDP, we need to truncate the request, clamp EDNS, and return it back. We don't for TCP.

Maybe we can avoid forwarding over TCP entirely if the request was from UDP? we just return the truncated response and the client can retry over TCP.

twitchyliquid64 commented on net/dns/resolver/forwarder.go:
```diff
@@ -577,6 +601,89 @@ func (f *forwarder) sendUDP(ctx context.Context, fq *forwardQuery, rr resolverAn
+		tcpFam = "tcp6"
+	}
+
+	conn, err := f.dialer.SystemDial(ctx, tcpFam, ipp.String())
```
I'm not familiar with `SystemDial` - is it already context-aware or should we set some deadlines on the conn?

andrew-d replied:
Yeah, it's context-aware and calls `DialContext`, which properly aborts if the input context is canceled/times out 👍

andrew-d commented:
> Maybe we can avoid forwarding over TCP entirely if the request was from UDP? we just return the truncated response and the client can retry over TCP.

That's a good point, actually; I just plumbed the "family" through to this function so that we can distinguish whether it's a UDP or TCP query, and then avoid falling back to TCP for an input UDP query (i.e. return the truncated query). Thoughts?

twitchyliquid64 reviewed (approved):
One last thing, might be good to have an envknob for this so we can quickly turn it off if we find issues?

And then turn down the envknob once its stable? yeah/nah?

andrew-d commented:
Good call on that; I added an envknob that can disable this, along with a controlknob (cc @bradfitz for review since we changed that recently). I figure we may want the ability to turn this off if it turns out to be buggy in some configurations; we can remove this knob after we're confident that things work properly.

andrew-d commented:
(rebased on top of the merge conflict in #9288)

twitchyliquid64 reviewed (approved):
LGTM.

@maisem as we need a stamp for the new controlknob, and Brad is out.

maisem commented on net/dns/resolver/forwarder.go:
```diff
@@ -601,6 +730,7 @@ func (f *forwarder) resolvers(domain dnsname.FQDN) []resolverAndDelay {
 type forwardQuery struct {
 	txid   txid
 	packet []byte
+	family string // "tcp" or "udp"
```
protocol

maisem replied:
family is also fine

maisem commented on net/dns/resolver/forwarder_test.go:
```diff
@@ -240,3 +250,203 @@ func FuzzClampEDNSSize(f *testing.F) {
+	}
+	if tcpLn == nil || udpLn == nil {
+		tb.Fatal("failed to listen on same port for TCP/UDP")
+	}
```
this seems like it'll be super flaky

maisem replied:
you also leak one listener

andrew-d replied:
FWIW, I ran this test with `gostress -count 5000` and it didn't flake once, so I can't say for sure that it's perfect, but it's hopefully not too bad? This just tries to get the same port, not any specific port, and lets the system assign a random one for TCP (which I'm assuming will be more contended).

maisem replied:
I think it depends on how happy the CI machines :)

andrew-d replied:
Done

maisem commented on net/dns/resolver/forwarder_test.go:
```diff
@@ -240,3 +250,203 @@ func FuzzClampEDNSSize(f *testing.F) {
+		}
+	}
+
+	go func() {
```
track these with a waitgroup and wait on it in cleanup?

andrew-d commented:
@maisem PTAL? Rebased on top of the `NodeCapMap` changes.

maisem commented on net/dns/resolver/forwarder.go:
```diff
@@ -577,6 +623,89 @@ func (f *forwarder) sendUDP(ctx context.Context, fq *forwardQuery, rr resolverAn
+		tcpFam = "tcp6"
+	}
+
+	conn, err := f.dialer.SystemDial(ctx, tcpFam, ipp.String())
```
should we attempt to keep these around or are they a use once connection?

andrew-d replied:
They can be re-used, but some DNS servers will close the connection after a single query so I didn't really want to deal with detecting that, handling cache sizing/eviction, etc.

andrew-d replied:
How so? If we hit this codepath, we just sent a UDP packet and got a truncated response so the device radio should already be active. I'm not super familiar with how battery management works; is a TCP connection different than UDP?

maisem replied:
just thinking about the tcp handshake + retry loops

andrew-d commented:
@maisem Okay, made the changes to de-flake the "listen on the same port" part of the test; PTAL?

maisem commented on net/dns/resolver/forwarder.go:
```diff
@@ -477,10 +484,49 @@ func (f *forwarder) send(ctx context.Context, fq *forwardQuery, rr resolverAndDe
+		// falling back to TCP for a truncated query if we can't return
+		// the results to the client.
+		return ret, nil
+	} else if skipTCPRetry() || (f.controlKnobs != nil && f.controlKnobs.DisableDNSForwarderTCPRetries.Load()) {
```
you can drop the elses

maisem commented on net/dns/resolver/forwarder.go:
```diff
@@ -477,10 +484,49 @@ func (f *forwarder) send(ctx context.Context, fq *forwardQuery, rr resolverAndDe
+
+	// Retry over TCP, best-effort; return the truncated UDP response if we
+	// cannot query via TCP.
+	ret2, err2 := f.sendTCP(ctx, fq, rr)
+	if err2 == nil {
```
```
if ret2, err2 := f.sendTCP(ctx, fq, rr); err2 == nil {
}
```

maisem commented on net/dns/resolver/forwarder.go:
```diff
@@ -577,6 +623,89 @@ func (f *forwarder) sendUDP(ctx context.Context, fq *forwardQuery, rr resolverAn
+		tcpFam = "tcp6"
+	}
+
+	conn, err := f.dialer.SystemDial(ctx, tcpFam, ipp.String())
```
should add a shorter timeout on the context

andrew-d replied:
Added

maisem commented on net/dns/resolver/forwarder_test.go:
```diff
@@ -240,3 +251,221 @@ func FuzzClampEDNSSize(f *testing.F) {
+}
+
+func TestForwarderTCPFallback(t *testing.T) {
+	envknob.Setenv("TS_DEBUG_DNS_FORWARD_SEND", "true")
```
unset after test?

andrew-d replied:
Good catch; had it read and restore the prior value after the test 👍

andrew-d commented:
@maisem - addressed all changes; PTAL?
